Restrict Variables subitems to Tags::Variables

nilsvu opened the pull request:
## Proposed changes

Tags other than `::Tags::Variables` will not automatically expose their `Variables` constituents as subitems to avoid inflating the size of the DataBox. Closes #3229.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

kidder commented on src/DataStructures/Variables.hpp:
```diff
@@ -841,8 +844,21 @@ struct NumberOfPoints<Variables<TagList>> {
+}  // namespace Variables_detail
 template <typename Tag>
-struct Subitems<Tag, Requires<tt::is_a_v<Variables, typename Tag::type>>> {
+struct Subitems<Tag, Requires<Variables_detail::is_a_variables_tag_v<Tag>>> {
```
why do you need the above; isn't this just `tt:is_a_v<::Tags::Variables, Tag>` ?

nilsvu replied:
No, this includes both the `Tags::Variables` template _and_ classes that inherit from the `Tags::Variables` template. It's needed for compute items inheriting from `Tags::Variables`.

nilsdeppe replied:
Since both @kidder  and I were confused, maybe add a comment why this isn't using `is_a_v` or `is_base_of`?

nilsvu replied:
sure, done

nilsvu commented on src/Evolution/DgSubcell/Tags/Inactive.hpp:
```diff
@@ -27,7 +27,8 @@ struct Inactive : db::PrefixTag, db::SimpleTag {
 template <typename TagList>
-struct Inactive<::Tags::Variables<TagList>> : db::PrefixTag, db::SimpleTag {
+struct Inactive<::Tags::Variables<TagList>>
+    : ::Tags::Variables<db::wrap_tags_in<Inactive, TagList>> {
```
@nilsdeppe this tag doesn't play nice with the change in this PR because it works differently than all other prefix tags: Other prefix tags rely on `db::add_tag_prefix` for the `Variables` specialization, which produces: `Tags::Variables<tmpl::list<Inactive<TensorTag>>>`. Note the result is a `Tags::Variables` with prefixed _constituents_, i.e. the prefix collapses to the inner tag list. In other words, it does _not_ produce: `Inactive<Tags::Variables<tmpl::list<Inactive<TensorTag>>>`.

I understand that it's annoying having to write `db::add_tag_prefix` everywhere, and I have wished to get rid of it for a long time. Do you have an idea how we could handle these Variables prefixes consistently, and also resolve #3229? This change here works ok, but I'm not particularly happy about it.

nilsdeppe replied:
I don't know... This looks like it should be okay. I think the `OnSubcells` can just be used with the `db::add_tag_prefix` instead. The `Inactive` gets used a lot more at the `Variables` even in user-facing code. They could probably both be changed over, but I don't really like the any of anything changing inside subcell code right now because it could quickly result in me having to debug something for a day, which there really isn't time for right now, unfortunately.

You can try sticking this diff on my `compute_dg_time_derivative` branch and if `EvolveValanciaDivCleanMagnetizedTovStar` compiles it should be fine, but even there I'm not sure everything that branch needs to compile has been pushed...

wthrowe replied:
See also #1123.

nilsvu replied:
I think the alternative is to add a static member variable like `bool add_variables_subitems` that, when present and true, enables the subitems. However, I don't really like to add such a "magic switch" to tags so I thought it's better to bake the subitems feature into `Tags::Variables`, which brings us to the current implementation. @kidder Please let me know when you have thought sufficiently carefully about this and reached a conclusion :)

nilsdeppe replied:
All of the relevant subcell code is merged, so switching to some other interface (like not having the prefixed Variables tags be prefixed themselves, but only the inner tags) would not cause any possibly very difficult merge conflicts.

nilsvu replied:
I left this as a possible compile-time optimization and added a comment. For now, no subcell behavior changed. There are more places in the code where we may want to omit the subitems to lighten the DataBox. For example, in the parallel linear solvers many subitems are added to the box but never used.

nilsdeppe replied:
Yea, that's fair :) I think all subcell code would be just fine requiring people to work with the full Variables instead of the individual tensors when dealing with inactive quantities :)

kidder commented on src/Evolution/DgSubcell/Tags/Inactive.hpp:
```diff
@@ -27,7 +27,8 @@ struct Inactive : db::PrefixTag, db::SimpleTag {
 template <typename TagList>
-struct Inactive<::Tags::Variables<TagList>> : db::PrefixTag, db::SimpleTag {
+struct Inactive<::Tags::Variables<TagList>>
+    : ::Tags::Variables<db::wrap_tags_in<Inactive, TagList>> {
```
requesting a change so this doesn't get merged before I have a chance to think about it carefully...

nilsvu committed:
Restrict Variables subitems to Tags::Variables

Other tags with Variables type will not automatically expose the
constituents as subitems to avoid inflating the size of the DataBox.

nilsvu commented:
@kidder Please take a look at this one

nilsvu commented:
Rebased on #3476. This PR reduces the number of tags in the DataBox for the `SolveXcts` executable from 243 to 219 right away, with no functional changes (the eliminated subitems were never used). I expect the number of tags can be reduced a lot further with some optimizations of tags in the linear solvers etc.

kidder commented:
@wthrowe @nilsdeppe do either of you want to look a this?  Otherwise it can be merged as far as I am concerned.
